primitives/state-machine/src/trie_backend_essence.rs

```diff
@@ -64,6 +64,76 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
 		self.storage
 	}
 
+	/// Return the next key in the trie i.e. the minimum key that is strictly superior to `key` in
+	/// lexicographic order.
+	pub fn next_storage_key(&self, key: &[u8]) -> Result<Option<Vec<u8>>, String> {
+		self.next_storage_key_from_root(&self.root, key)
+	}
+
+	/// Return the next key in the child trie i.e. the minimum key that is strictly superior to
+	/// `key` in lexicographic order.
+	pub fn next_child_storage_key(
+		&self,
+		storage_key: &[u8],
+		key: &[u8],
+	) -> Result<Option<Vec<u8>>, String> {
+		let child_root = match self.storage(storage_key)? {
+			Some(child_root) => child_root,
+			None => return Ok(None),
+		};
+
+		let mut hash = H::Out::default();
+
+		if child_root.len() != hash.as_ref().len() {
+			return Err(format!("Invalid child storage hash at {:?}", storage_key));
+		}
+		// note: child_root and hash must be same size, panics otherwise.
+		hash.as_mut().copy_from_slice(&child_root[..]);
+
+		self.next_storage_key_from_root(&hash, key)
+	}
+
+	/// Return next key from main trie or child trie by providing corresponding root.
+	fn next_storage_key_from_root(
+		&self,
+		root: &H::Out,
+		key: &[u8],
+	) -> Result<Option<Vec<u8>>, String> {
+		let mut read_overlay = S::Overlay::default();
+		let eph = Ephemeral {
+			storage: &self.storage,
+			overlay: &mut read_overlay,
+		};
+
+		let trie = TrieDB::<H>::new(&eph, root)
+			.map_err(|e| format!("TrieDB creation error: {}", e))?;
+		let mut iter = trie.iter()
+			.map_err(|e| format!("TrieDB iteration error: {}", e))?;
+
+		// The key just after the one given in input, basically `key++0`.
+		// Note: We are sure this is the next key if:
+		// * size of key has no limit (i.e. we can always add 0 to the path),
+		// * and no keys can be inserted between `key` and `key++0` (this is ensured by sr-io).
```

Comment:
What does this mean and how is that ensured?

Reply:
we doesn't want to iterate from the key included but from the key excluded and trie crate only provides seek function.

So we compute the next potential key, by doing next_potential_key = key++0.

But this is only true if:
* we can indeed add 0 to the key
* and if there is no key in the trie between key and key++0

First option requires size of key has no limit or that the runtime is not suppose to write anything near this limit, and the second option is ensured because sr-io doesn't provide any way to write between key and key++0, even if it is a 16-trie.

Maybe @cheme you can confirm this ?

But maybe it is better to actually don't make those assumption and seek from the key, get the next, if the next is the current one then get the second next.

Reply:
that would result in something like this:
```diff
diff --git a/primitives/state-machine/src/trie_backend_essence.rs b/primitives/state-machine/src/trie_backend_essence.rs
index aea0193dd..1c8bafb09 100644
--- a/primitives/state-machine/src/trie_backend_essence.rs
+++ b/primitives/state-machine/src/trie_backend_essence.rs
@@ -110,23 +110,29 @@ impl<S: TrieBackendStorage<H>, H: Hasher> TrieBackendEssence<S, H> where H::Out:
                let mut iter = trie.iter()
                        .map_err(|e| format!("TrieDB iteration error: {}", e))?;
 
-               // The key just after the one given in input, basically `key++0`.
-               // Note: We are sure this is the next key if:
-               // * size of key has no limit (i.e. we can always add 0 to the path),
-               // * and no keys can be inserted between `key` and `key++0` (this is ensured by sr-io).
-               let mut potential_next_key = Vec::with_capacity(key.len() + 1);
-               potential_next_key.extend_from_slice(key);
-               potential_next_key.push(0);
-
-               iter.seek(&potential_next_key)
+               iter.seek(&key)
                        .map_err(|e| format!("TrieDB iterator seek error: {}", e))?;
 
                let next_element = iter.next();
 
                let next_key = if let Some(next_element) = next_element {
-                       let (next_key, _) = next_element
-                               .map_err(|e| format!("TrieDB iterator next error: {}", e))?;
-                       Some(next_key)
+                       let next_key = next_element
+                               .map_err(|e| format!("TrieDB iterator next error: {}", e))?.0;
+
+                       if next_key != key {
+                               Some(next_key)
+                       // If next key is same as key iter to second next element.
+                       } else {
+                               let next_element = iter.next();
+                               if let Some(next_element) = next_element {
+                                       let next_key = next_element
+                                               .map_err(|e| format!("TrieDB iterator next error: {}", e))?.0;
+
+                                       Some(next_key)
+                               } else {
+                                       None
+                               }
+                       }
                } else {
                        None
                };
```

Reply:
Yes, the fact that api only allow to write key value at byte address is important here (otherwhise we would need to work with nibbles (half byte for 16-trie) and add 0 nibble.
I am not sure if skipping the first iteration is very important here (we need to query the node anyway during seek call).
The relevant perf gain would be from using the same iterator between calls (to skip the 'seek' call on every 'next), but it is not a easy todo.

Reply:
Should we not rely on "the fact that api only allow to write key value at byte address", this property could be leverage in the future.